Add JSON output to tap command

kleimkuhler opened the pull request:
Replaces #3411 

### Motivation

It is a little tough to filter/read the current tap output. As headers are being
added to tap, the output is starting to get difficult to consume. Take a peek at
#3262 for an example. It would be nice to have some more machine readable output
that can be sliced and diced with tools such as jq.

### Solution

A new output option has been added to the `linkerd tap` command that returns the
JSON encoding of tap events.

The default output is line oriented; `-o wide` appends the request's target
resource type to the tap line oriented tap events.

In order display certain values in a more human readable form, a tap event
display struct has been introduced. This struct maps public API `TapEvent`s
directly to a private `tapEvent`. This struct offers a flatter JSON structure
than the protobuf JSON rendering. It also can format certain field--such as
addresses--better than the JSON protobuf marshaler.

Closes #3390

**Default**:
```
➜  linkerd2 git:(kleimkuhler/tap-json-output) linkerd -n linkerd tap deploy/linkerd-web
req id=5:0 proxy=in  src=10.1.6.146:36976 dst=10.1.6.148:9994 tls=not_provided_by_remote :method=GET :authority=10.1.6.148:9994 :path=/metrics
rsp id=5:0 proxy=in  src=10.1.6.146:36976 dst=10.1.6.148:9994 tls=not_provided_by_remote :status=200 latency=3366µs
end id=5:0 proxy=in  src=10.1.6.146:36976 dst=10.1.6.148:9994 tls=not_provided_by_remote duration=132µs response-length=1505B
```

**Wide**:
```
➜  linkerd2 git:(kleimkuhler/tap-json-output) linkerd -n linkerd tap deploy/linkerd-web -o wide
req id=6:0 proxy=in  src=10.1.0.1:35394 dst=10.1.6.148:9994 tls=not_provided_by_remote :method=GET :authority=10.1.6.148:9994 :path=/ping dst_res=deploy/linkerd-web dst_ns=linkerd
rsp id=6:0 proxy=in  src=10.1.0.1:35394 dst=10.1.6.148:9994 tls=not_provided_by_remote :status=200 latency=1442µs dst_res=deploy/linkerd-web dst_ns=linkerd
end id=6:0 proxy=in  src=10.1.0.1:35394 dst=10.1.6.148:9994 tls=not_provided_by_remote duration=88µs response-length=5B dst_res=deploy/linkerd-web dst_ns=linkerd
```

**JSON**:
*Edit: Flattened `Method` and `Scheme` formatting*
```
{
  "source": {
    "ip": "10.138.0.28",
    "port": 47078,
    "metadata": {
      "daemonset": "ip-masq-agent",
      "namespace": "kube-system",
      "pod": "ip-masq-agent-4d5s9",
      "serviceaccount": "ip-masq-agent",
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.60.1.49",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-6988999458-c6wpw",
      "pod_template_hash": "6988999458",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "requestInitEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "method": "GET",
    "scheme": "",
    "authority": "10.60.1.49:9994",
    "path": "/ready"
  }
}
{
  "source": {
    "ip": "10.138.0.28",
    "port": 47078,
    "metadata": {
      "daemonset": "calico-node",
      "namespace": "kube-system",
      "pod": "calico-node-bbrjq",
      "serviceaccount": "calico-sa",
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.60.1.49",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-6988999458-c6wpw",
      "pod_template_hash": "6988999458",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "responseInitEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "sinceRequestInit": {
      "nanos": 644820
    },
    "httpStatus": 200
  }
}
{
  "source": {
    "ip": "10.138.0.28",
    "port": 47078,
    "metadata": {
      "deployment": "calico-typha",
      "namespace": "kube-system",
      "pod": "calico-typha-59cb487c49-8247r",
      "pod_template_hash": "59cb487c49",
      "serviceaccount": "calico-sa",
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.60.1.49",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-6988999458-c6wpw",
      "pod_template_hash": "6988999458",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "responseEndEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "sinceRequestInit": {
      "nanos": 790898
    },
    "sinceResponseInit": {
      "nanos": 146078
    },
    "responseBytes": 3,
    "grpcStatusCode": 0
  }
}
```

Signed-off-by: Kevin Leimkuhler <kleimkuhler@icloud.com>

kleimkuhler commented on pkg/addr/addr.go:
```diff
@@ -14,9 +16,25 @@ import (
+// If Ipv6, the bytes should be ordered big-endian. When formatted as a
+// string, the IP address should be enclosed in square brackets followed by
+// the port.
 func PublicAddressToString(addr *public.TcpAddress) string {
```
This function has been updated to format both Ipv4 and Ipv6 IP addresses. When Ipv6, the IP address is enclosed in square brackets to separate it from the port.

I changed the event destination to be IPv6 and updated the tests accordingly.

I could separate the IPv4/IPv6 formatting changes from this PR, but the changes were related to handling IPv6 possibility in `TapEvent`s and are confined to these tests only.

kleimkuhler commented:
The main difference between this and #3411 is that it is not based off of tap headers being present in `TapEvent`. It also formats JSON fields to be more human readable than we could achieve with the JSON protobuf marshaler.

grampelberg commented:
Looking pretty great to me. WDYT about splitting the port out from the ip?

kleimkuhler commented:
@grampelberg Yep good call!

adleong commented on cli/cmd/tap.go:
```diff
@@ -31,6 +35,51 @@ type tapOptions struct {
+
+type requestInitEvent struct {
+	ID        *streamID      `json:"id"`
+	Method    *pb.HttpMethod `json:"method"`
```
this struct is an enum and doesnt render well.  lets make this a string field like "GET" or "POST"

adleong commented on cli/cmd/tap.go:
```diff
@@ -292,6 +348,49 @@ func renderTapEvent(event *pb.TapEvent, resource string) string {
+	// Map destination endpoint
+	dip := addr.PublicIPToString(event.GetDestination().GetIp())
+	d := &endpoint{
+		IP:       dip,
```
oh dip!

kleimkuhler commented on cli/cmd/tap.go:
```diff
@@ -31,6 +35,51 @@ type tapOptions struct {
+	SinceRequestInit  *duration.Duration `json:"sinceRequestInit"`
+	SinceResponseInit *duration.Duration `json:"sinceResponseInit"`
+	ResponseBytes     uint64             `json:"responseBytes"`
+	GrpcStatusCode    uint32             `json:"grpcStatusCode,omitempty"`
```
Oh okay good call. I'll remove the omitempty

kleimkuhler commented:
@adleong I made the requested changes. I reordered some functions around so that it reads without jumping around as much--apologies if that makes the change diff big for you again.

`Scheme` in RequestInit does render as `""` because it doesn't appear to be set properly. I added to the tests so that when it is there it checks that it renders properly, but that does not appear to be the case in actual output. I'll open an issue to investigate that further.

I updated the example output in the PR description!

siggy commented on cli/cmd/tap.go:
```diff
@@ -140,38 +191,43 @@ func newCmdTap() *cobra.Command {
-	tableWriter := tabwriter.NewWriter(w, 0, 0, 0, ' ', tabwriter.AlignRight)
-	err := writeTapEventsToBuffer(tapByteStream, tableWriter, resource)
+func writeTapEventsToBuffer(w io.Writer, tapByteStream *bufio.Reader, req *pb.TapByResourceRequest, options *tapOptions) error {
+	writer := tabwriter.NewWriter(w, 0, 0, 0, ' ', tabwriter.AlignRight)
```
`tabwriter` doesn't make much sense for json output. also the default and wide outputs are not outputting delimited output either, so we should be safe to remove it:
```diff
 func writeTapEventsToBuffer(w io.Writer, tapByteStream *bufio.Reader, req *pb.TapByResourceRequest, options *tapOptions) error {
-       writer := tabwriter.NewWriter(w, 0, 0, 0, ' ', tabwriter.AlignRight)
-
        var err error
        switch options.output {
        case "":
-               err = renderTapEvents(tapByteStream, writer, renderTapEvent, "")
+               err = renderTapEvents(tapByteStream, w, renderTapEvent, "")
        case wideOutput:
                resource := req.GetTarget().GetResource().GetType()
-               err = renderTapEvents(tapByteStream, writer, renderTapEvent, resource)
+               err = renderTapEvents(tapByteStream, w, renderTapEvent, resource)
        case jsonOutput:
-               err = renderTapEvents(tapByteStream, writer, renderTapEventJSON, "")
+               err = renderTapEvents(tapByteStream, w, renderTapEventJSON, "")
        }
        if err != nil {
                return err
        }
-       writer.Flush()

        return nil
 }

-func renderTapEvents(tapByteStream *bufio.Reader, w *tabwriter.Writer, render renderTapEventFunc, resource string) error {
+func renderTapEvents(tapByteStream *bufio.Reader, w io.Writer, render renderTapEventFunc, resource string) error {
```

siggy commented on cli/cmd/tap.go:
```diff
@@ -292,6 +348,121 @@ func renderTapEvent(event *pb.TapEvent, resource string) string {
+	m := mapPublicToDisplayTapEvent(event)
+	e, err := json.MarshalIndent(m, "", "  ")
+	if err != nil {
+		return fmt.Sprintf("Error marshalling JSON: %s\n", err)
```
maybe this should render the error in json format, so as not to break readers?

siggy commented on cli/cmd/tap.go:
```diff
@@ -292,6 +348,121 @@ func renderTapEvent(event *pb.TapEvent, resource string) string {
+func getRequestInitEvent(pubEv *pb.TapEvent_Http) *requestInitEvent {
+	reqI := pubEv.GetRequestInit()
+	if reqI != nil {
+		sid := &streamID{
```
consider:
```go
reqI := pubEv.GetRequestInit()
if reqI == nil {
	return nil
}
sid := &streamID{
...
```
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

siggy commented on cli/cmd/tap_test.go:
```diff
@@ -92,16 +102,22 @@ func busyTest(t *testing.T, wide bool) {
 		goldenFilePath = "testdata/tap_busy_output_wide.golden"
-	} else {
+	case jsonOutput:
+		goldenFilePath = "testdata/tap_busy_output_json.golden"
```
🙏

kleimkuhler commented on go.mod:
```diff
@@ -29,6 +29,7 @@ require (
 	github.com/go-openapi/spec v0.17.2
 	github.com/go-openapi/swag v0.17.2 // indirect
 	github.com/gobwas/glob v0.2.3 // indirect
+	github.com/gogo/protobuf v1.3.0
```
Thanks this was from previously using the protobuf marshaler and not removing it!

kleimkuhler commented:
@siggy Changes made--thanks those were helpful! The error string is in valid JSON format as well so it should render okay if that branch is hit.
